### DISCUSSION TITLE
Inquiry about setting up a Milvus replication cluster using Milvus-CDC

### DISCUSSION THREAD

**Discussion by ian-lss** [milvus-io/milvus#33996] [Q&A and General discussion]
Hello,

I am trying to set up a replication milvus cluster using Milvus-CDC. 
I have confirmed that incremental data replication works well as stated in the documentation.

> As of now, Milvus-CDC is designed to synchronize only incremental data.

To set up a replication database similar to MySQL, 
I need to back up the previous data using Milvus-backup, restore this backup data to the replica Milvus cluster, and then use the recorded position from the backup to register Milvus-CDC. 
However, the latest version of Milvus-backup does not record this position.
Should I use the [flush segments] information in the milvus-backup log?

How can I obtain this position(checkpoint?)?

**Comment by xiaofan-luan** [comment]
@SimFG could help on this

**Comment by SimFG** [comment]
@ian-lss 
Currently, the linkage between backup tool and cdc tool is under development, and it is expected to be completed in mid-July.

 If you need to use it now, it cannot support the given position. The current usage mode: stop writing data to milvus, then use the backup tool to back up the existing data, restore this part of the data to another milvus, and then start cdc synchronization without transferring the point. After cdc works normally, milvus can be used normally.

In subsequent versions, the backup tool supports returning the position, and the cdc tool supports passing the position. At this time, the stop time of source milvus only needs to get the backup tool return position to work normally.

**Comment by ian-lss** [comment]
@SimFG @xiaofan-luan 

I really appreciate both of your responses, and I will be looking forward to mid-July.

Thank you.

**Comment by Uijeong97** [comment]
@SimFG @xiaofan-luan

>Currently, the linkage between backup tool and cdc tool is under development, and it is expected to be completed in mid-July.

I am wondering if the Backup-cdc integration has been implemented.
Can you let me know?

**Comment by SimFG** [comment]
@Uijeong97 This function has been implemented in the latest version of cdc. You can try to add the position parameter in the collection info parameter in the cdc create request. For example:

```
{
  "collection_infos": [
    {
      "name": "hello_milvus",
      "positions": {
        "by-dev-rootcoord-dml_0_449777144535976857v0": "xxxxxxx"
      }
    }
  ]
}
```

However, the cdc function is currently under testing, and some other functions are under development. There is currently no documentation for this part.

**Reply by ian-lss** [reply]
@SimFG 
Hi,

Based on the above content, I tried entering what appears to be the Backup log’s position and running it, but there hasn’t been much change from the previous situation. That is, after restoring a backup made early in the morning, I reconnected the CDC 15 hours later, but only the data after the CDC connection was replicated.

Below is the information from the milvus-backup log.
Isn’t the value that should be entered in place of "by-dev-rootcoord-dml_0_449777144535976857v0": "xxxxxxx" the one from channel_checkpoints?

If the values are being used correctly, what aspects should be checked?

All Milvus solutions are the latest version.

[2024/09/08 18:00:31.712 +00:00] [INFO] [core/backup_impl_create_backup.go:515] [backupCollectionExecute] [collectionMeta="id:\"xxxxxxx\" start_time:1725818431 collection_id:000000000 db_name:\"cdctest\" collection_name:\"replica_collection\" schema:<name:\"replica_collection\" description:\"replica test\" fields:<fieldID:100 name:\"id\" is_primary_key:true data_type:Int64 autoID:true > fields:<fieldID:101 name:\"vector\" data_type:FloatVector type_params:<key:\"dim\" value:\"1563\" > > > shards_num:1 consistency_level:Bounded backup_timestamp:452412933747507201 has_index:true index_infos:<field_name:\"vector\" index_name:\"iv_vector\" index_type:\"HNSW\" params:<key:\"index_type\" value:\"HNSW\" > params:<key:\"metric_type\" value:\"L2\" > params:<key:\"params\" value:\"{\\\"M\\\":20,\\\"efConstruction\\\":200}\" > > load_state:\"NotLoad\" backup_physical_timestamp:1725818431 channel_checkpoints:<key:\"dbe-rootcoord-dml_0_452360423430695105v0\" value:\"CihkYmUtcm9vdGNvb3JkLWRtbF8wXzQ1MjM2MDQyMzQzMDY5NTEwNXYwEgiVClEAAAAAABo6ZGF0YW5vZGUtMjU4LWRiZS1yb290Y29vcmQtZG1sXzBfNDUyMzYwNDIzND=\" > "]

**Reply by SimFG** [reply]
> Isn’t the value that should be entered in place of "by-dev-rootcoord-dml_0_449777144535976857v0": "xxxxxxx" the one from channel_checkpoints?

yes.

> If the values are being used correctly, what aspects should be checked?

when the backup process is over, it should return the channel checkpoints. Is there no return value after the backup is completed?

**Reply by ian-lss** [reply]
@SimFG 
I am so sorry. 
It seems I mistakenly copied the position information that was logged. After completing the backup, I used the checkpoint information from the final response, and it worked correctly. 
Currently, only one collection position can be registered per task, but do you have plans to allow registering all collection positions under a single task?

**Reply by SimFG** [reply]
There will be plans to support it later, but we are currently dealing with some issues related to cdc usage restrictions.

**Reply by rere950303** [reply]
@SimFG 

When the backup is complete, the `channel_cp_meta.json` metadata file contains a checkpoint for each collection's pchannel, as shown below.

```json
{
  "docs_V1_FILE": [
    {
      "name": "by-dev-rootcoord-dml_13",
      "position": "CixieS1kZXYtcm9vdGNvb3JkLWRtbF8xM180NTQ1MDY2MjgyODc0NDIzMTF2MBILCI4/EOjpAhgAIAAaPmRhdGFub2RlLTU3Ni1ieS1kZXYtcm9vdGNvb3JkLWRtbF8xM180NTQ1MDY2MjgyODc0NDIzMTF2MC10cnVlIIKAkIW/9PKpBg=="
    }
  ],
  ...
}
```

In `collection_infos`, can i set it to `pchannel` instead of a `vchannel` like `by-dev-rootcoord-dml_0_449777144535976857v0`?

```json
{
  "collection_infos": [
    {
      "name": "docs_V1_FILE",
      "positions": {
        "by-dev-rootcoord-dml_13": "xxxxxxx"
      }
    }
  ]
}
```

**Reply by huanghaoyuanhhy** [reply]
@rere950303 The Backup tool converts vch to pch, which is actually unnecessary and will be fixed.

**Reply by huanghaoyuanhhy** [reply]
https://github.com/zilliztech/milvus-backup/issues/543

**Reply by rere950303** [reply]
@huanghaoyuanhhy 
Thank you. Then, I guess we'll have to use only VCH at the CDC.
